guava/src/com/google/common/io/ReaderInputStream.java

```diff
@@ -185,6 +185,76 @@ public int read(byte[] b, int off, int len) throws IOException {
     }
   }
 
+  @Override
+  public int available() throws IOException {
+    if (byteBuffer.hasRemaining()) {
+      // if we have remaining chars in the buffer, return their count
+      return byteBuffer.remaining();
+    } else if (charBuffer.hasRemaining()) {
+      // if not, try to make quick progress by encoding available
+      // chars that have been read from the users reader
+      CoderResult result = encoder.encode(charBuffer, byteBuffer, endOfInput);
+
+      if (result.isOverflow()) {
+        // the byte buffer is too small, enlarge it
+        startDraining(true);
+      } else if (result.isUnderflow()) {
+        // the char buffer is too small, enlarge it
+        charBuffer = grow(charBuffer);
+      } else if (result.isError()) {
+        // Only reach here if a CharsetEncoder with non-REPLACE settings is used.
+        result.throwException();
+      }
+      return available();
+
+    } else if (!endOfInput && reader.ready()) {
+      // if none of the above was true, check if
+      // the reader is ready, and try to make progress on it
+      int output;
+
+      // the flip is necessary for appending at the right position
+      Java8Compatibility.flip(charBuffer);
+      // read until the read operation would block
+      while (reader.ready()) {
```

Comment:
Though as additional optimization you could check whether the class of the reader is known to never block, i.e. [`StringReader`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/StringReader.html) (but to be safe check `getClass() ==` instead of `instanceof` because `StringReader` makes no guarantees whether it blocks and maybe someone wrote a blocking subclass) and Guava's [`CharSequenceReader`](https://github.com/google/guava/blob/master/guava/src/com/google/common/io/CharSequenceReader.java).

Reply:
> ath.min(charBuffer.remaining(), X)

Yeah that's a good question. Implementing the `Math.min(charBuffer.remaining(), X)` approach could lead to `available()` being blocked... on the other hand reading char by char is horribly inefficient.

Reply:
Ah, sorry my comment was not that clear; I meant that the `Math.min(charBuffer.remaining(), X)` limit should be used as additional condition for the `read()` loop, i.e. reading a single char at a time. This is still pretty inefficient, but at least not as inefficient as filling the complete buffer, especially if the caller might not even be interested in that many bytes.